components/places/src/api/places_api.rs

```diff
@@ -240,39 +287,37 @@ impl PlacesApi {
         }
 
         let sync_state = guard.as_ref().unwrap();
-        // Note that counter-intuitively, this must be called before we do a
-        // bookmark sync too, to ensure the shared global state is correct.
+        // Note that this *must* be called before either history or bookmarks are
+        // synced, to ensure the shared global state is correct.
         HistoryStore::migrate_v1_global_state(&conn)?;
 
-        let interruptee = conn.begin_interrupt_scope();
-        let store = HistoryStore::new(&conn, &interruptee);
         let mut mem_cached_state = sync_state.mem_cached_state.take();
         let mut disk_cached_state = sync_state.disk_cached_state.take();
-        let mut sync_ping = telemetry::SyncTelemetryPing::new();
-        let result = store.sync(
-            &client_init,
-            &key_bundle,
-            &mut mem_cached_state,
-            &mut disk_cached_state,
-            &mut sync_ping,
-        );
+        let mut result = syncer(&conn, &mut mem_cached_state, &mut disk_cached_state);
         // even on failure we set the persisted state - sync itself takes care
         // to ensure this has been None'd out if necessary.
         self.set_disk_persisted_state(&conn, &disk_cached_state)?;
         sync_state.mem_cached_state.replace(mem_cached_state);
         sync_state.disk_cached_state.replace(disk_cached_state);
 
-        result?;
-
-        Ok(sync_ping)
+        // for b/w compat reasons, we do some dances with the result.
+        if let Err(e) = result.result {
+            return Err(e.into());
+        }
+        match result.engine_results.remove(name) {
+            None | Some(Ok(())) => Ok(result.telemetry),
+            Some(Err(e)) => Err(e.into()),
+        }
     }
 
-    // TODO: reduce duplication with above
-    pub fn sync_bookmarks(
+    // This is the new sync API until the sync manager lands. It's currently
+    // not wired up via the FFI - it's possible we'll do declined engines too
+    // before we do.
+    pub fn sync(
         &self,
         client_init: &sync15::Sync15StorageClientInit,
         key_bundle: &sync15::KeyBundle,
-    ) -> Result<telemetry::SyncTelemetryPing> {
+    ) -> Result<SyncResult> {
```

Comment:
Hmm, what do you think about making this return `SyncResult`, and not `Result<SyncResult>`? IIUC, this will discard the service status, partial telemetry for failed syncs, and which engines succeeded or failed (if, say, we fail to update the global state), and require callers to handle both this error, and errors in `SyncResult.result`. Is that right?

Or...if we want to keep the `Result`, should we make a richer error type that also carries the `ServiceStatus` and `SyncTelemetryPing`? Or have this return `Result<SyncResult, SyncResult>`?

Reply:
> Hmm, what do you think about making this return `SyncResult`

Yes, it should - I neglected that as there are currently zero callers so basically forgot :) While it could easily be done in any followup which introduces a caller to that, I'll make that change now.